metrics-facade/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+use std::error;
+use std::fmt;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[macro_use]
+mod macros;
+
+static mut RECORDER: &'static MetricsRecorder = &NopRecorder;
+static STATE: AtomicUsize = AtomicUsize::new(0);
+
+const UNINITIALIZED: usize = 0;
+const INITIALIZING: usize = 1;
+const INITIALIZED: usize = 2;
+
+static SET_RECORDER_ERROR: &'static str = "attempted to set a recorder after the metrics system was already initialized";
+
+pub trait MetricsRecorder {
+    fn record_counter(&self, key: &str, value: u64);
+    fn record_gauge(&self, key: &str, value: i64);
+    fn record_histogram(&self, key: &str, value: u64);
+}
+
+struct NopRecorder;
+
+impl MetricsRecorder for NopRecorder {
+    fn record_counter(&self, _key: &str, _value: u64) { }
+    fn record_gauge(&self, _key: &str, _value: i64) { }
+    fn record_histogram(&self, _key: &str, _value: u64) { }
+}
+
+pub fn set_boxed_recorder(recorder: Box<MetricsRecorder>) -> Result<(), SetRecorderError> {
+    set_recorder_inner(|| unsafe { &*Box::into_raw(recorder) })
+}
+
+fn set_recorder_inner<F>(make_recorder: F) -> Result<(), SetRecorderError>
+where
+    F: FnOnce() -> &'static MetricsRecorder,
+{
+    unsafe {
```

Comment:
what is actually unsafe here...?

Reply:
Mutable static.

    error[E0133]: use of mutable static is unsafe and requires unsafe function or block
      --> metrics-facade/src/lib.rs:43:17
       |
    43 |                 RECORDER = make_recorder();
       |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ use of mutable static
       |
       = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

Reply:
ah....